pkg/node/manager/manager.go

```diff
@@ -532,3 +539,42 @@ func (m *Manager) DeleteAllNodes() {
 	m.nodes = map[nodeTypes.Identity]*nodeEntry{}
 	m.mutex.Unlock()
 }
+
+// StartNeighborRefresh spawns a controller which refreshes neighbor table
+// by sending arping periodically. Linux keeps an arp entry in reachable
+// state for (0.5 ~ 1.5) * base_reachable_time_ms, default by 15 to 45 seconds:
+// https://elixir.bootlin.com/linux/v5.7.19/source/net/core/neighbour.c#L113
+// We do the refresh in a similar way.
+func (m *Manager) StartNeighborRefresh(nh datapath.NodeHandler) {
+	baseNeighborRefreshInterval := option.Config.NeighborRefreshBaseInterval
+	if baseNeighborRefreshInterval <= 0 {
+		baseNeighborRefreshIntervalStr, err := sysctl.Read("net.ipv4.neigh.default.base_reachable_time_ms")
+		if err != nil {
+			baseNeighborRefreshInterval = 30000 * time.Millisecond
+		} else {
+			baseNeighborRefreshIntervalU32, err := strconv.ParseUint(baseNeighborRefreshIntervalStr, 10, 32)
+			if err != nil {
+				baseNeighborRefreshInterval = 30000 * time.Millisecond
+			} else {
+				baseNeighborRefreshInterval = time.Duration(baseNeighborRefreshIntervalU32) * time.Millisecond
+			}
+		}
+	}
+	controller.NewManager().UpdateController("neighbor-table-refresh",
+		controller.ControllerParams{
+			DoFunc: func(context.Context) error {
+				n := randGen.Int63n(int64(baseNeighborRefreshInterval))
```

Comment:
Is the neigh rand reach time in the kernel calculated for each entry or for a neigh GC run?

With the current approach we would generate a traffic burst when there are many nodes. Also, a recently added node (i.e. `now() - $NODE_ADD_TIME < $BASE_REFRESH_INTERVAL`) would unnecessary refreshed.

If we decide to calculate the refresh period of each node, then you could call `nh.NodeNeighborRefresh(...)` from a separate goroutine.

Reply:
It seems the state change is triggered by a per entry timer and reachable time is refreshed for each entry at periodic GC run. Since we don't have GC logic, what about setting controller interval to `1.5 * $BASE_REFRESH_INTERVAL` and
```
for _, entry := range m.nodes {
	go func() {
		sleep(0.5 *  $BASE_REFRESH_INTERVAL + random($BASE_REFRESH_INTERVAL))
		nh.NodeNeighborRefresh(entry.node)
	}()
}
```
This should ensure ARP probes are spread in the (0.5 ~ 1.5) * $BASE_REFRESH_INTERVAL time range

Reply:
With `sleep(0.5 * $BASE_REFRESH_INTERVAL + random($BASE_REFRESH_INTERVAL))` and controller interval `1.5 * $BASE_REFRESH_INTERVAL` I think that the refresh period for an entry will be `[2; 3] * $BASE_REFRESH_INTERVAL`. You need to set the controller interval to `$BASE_REFRESH_INTERVAL` in order for the range to be `[0.5; 1.5] * $BASE_REFRESH_INTERVAL`.

Anyway, such approach should work. Just one concern is that in the controller we don't wait for all goroutines to finish, so it might be that there will be multiple goroutines for the same node.

Another thing, have you considered making locking in `refreshNeighbor()` more fine grained (to make use of parallel arpings)?

Reply:
I think it should be [0.5; 1.5] since the sleep in goroutines and controller interval should start ticking at the same time(regardless of the time cost of refresh) IIUC.

> Just one concern is that in the controller we don't wait for all goroutines to finish, so it might be that there will be multiple goroutines for the same node.

Thanks, I'll add a wait logic.

> Another thing, have you considered making locking in `refreshNeighbor()` more fine grained (to make use of parallel arpings)?

Sorry, what do you mean by `parallel arpings`?

Reply:
> I think it should be [0.5; 1.5] since the sleep in goroutines and controller interval should start ticking at the same time(regardless of the time cost of refresh) IIUC.

AFAIU, the next execution of the controller will be scheduled after `1.5 * $BASE_REFRESH_INTERVAL` since the last return. Before arping can be performed in a goroutine, there will be a sleep. These two add up to the `[2; 3]`.

> Thanks, I'll add a wait logic.

Just looked into `pkg/controller/controller.go`, it seems that it doesn't take the execution duration into consideration when calculating next exec time. This means that the next exec time will be influenced by a goroutine which exec took the longest. I think it might be fine to avoid waiting, as in many cases `$BASE_REFRESH_INTERVAL >> arping time` which should guarantee that there won't be multiple goroutines for the same node.

>  I would prefer keeping refreshNeighbor separated and adjust locking in refreshNeighbor.

Sure. Also, we can do it as a follow up in order to get this PR merged sooner.

Reply:
Oops, yes, the right approach should be setting controller interval to 1 and sleep `[0.5; 1]` in goroutine and the result would be `[0.5; 1.5]`